x/wasm/keeper/relay.go

```diff
@@ -130,13 +133,34 @@ func (k Keeper) OnRecvPacket(
 	res, gasUsed, execErr := k.wasmVM.IBCPacketReceive(codeInfo.CodeHash, env, msg, prefixStore, cosmwasmAPI, querier, ctx.GasMeter(), gas, costJSONDeserialization)
 	k.consumeRuntimeGas(ctx, gasUsed)
 	if execErr != nil {
-		panic(execErr)
+		panic(execErr) // let contract fully abort IBC receive in certain case
 	}
-	if res.Err != "" { // handle error case as before https://github.com/CosmWasm/wasmvm/commit/c300106fe5c9426a495f8e10821e00a9330c56c6
-		return nil, sdkerrors.Wrap(types.ErrExecuteFailed, res.Err)
+	if res.Err != "" {
+		// return error ACK with non-redacted contract message, state will be reverted
+		return channeltypes.Acknowledgement{
+			Response: &channeltypes.Acknowledgement_Error{Error: res.Err},
+		}, nil
 	}
 	// note submessage reply results can overwrite the `Acknowledgement` data
-	return k.handleContractResponse(ctx, contractAddr, contractInfo.IBCPortID, res.Ok.Messages, res.Ok.Attributes, res.Ok.Acknowledgement, res.Ok.Events)
+	data, err := k.handleContractResponse(ctx, contractAddr, contractInfo.IBCPortID, res.Ok.Messages, res.Ok.Attributes, res.Ok.Acknowledgement, res.Ok.Events)
+	if err != nil {
+		// submessage errors result in error ACK with state reverted. Error message is redacted
+		return nil, err
+	}
+	// success ACK, state will be committed
+	return ContractConfirmStateAck(data), nil
+}
+
+var _ ibcexported.Acknowledgement = ContractConfirmStateAck{}
+
+type ContractConfirmStateAck []byte
+
+func (w ContractConfirmStateAck) Success() bool {
+	return true // always commit state
```

Comment:
I brought this up here: https://github.com/cosmos/ibc-go/pull/3515. Let's see if that provides clarity what this function means.

Thinking about it more, it seems like `ibcexported.Acknowledgement` is not an acknowledgement but a container around the acknowledgement.

Reply:
Ja, you can consider it a wrapper. The ACK data itself contains no information if the TX state should be persisted or not. So custom implementations add this "context" data by implementing the `Success()` method. The name is not great it could be more verbose like `IsStatePersistable()`.
For us this is always `true` when the contract "handled" the package without an error or Result.Err.